File: src/main/java/com/team4099/lib/vision/LogUtil.java
package com.team4099.lib.vision;

import edu.wpi.first.math.geometry.Pose2d;
import edu.wpi.first.math.geometry.Pose3d;

import java.util.List;

public final class LogUtil {
    public static double[] toPoseArray2d(Pose2d pose) {
        return new double[]{
            pose.getX(),
            pose.getY(),
            pose.getRotation().getRadians()
        };
    }
    public static double[] toPoseArray2d(List<Pose2d> poses) {
        double[] result = new double[3 * poses.size()];
        for(int i = 0; i < poses.size(); i++) {
            double[] vals = toPoseArray2d(poses.get(i));
            for(int j = 0; j < vals.length; j++) {
                result[3*i + j] = vals[j];
            }
        }
        return result;
    }
    public static double[] toPoseArray3d(Pose3d pose) {
        var rot = pose.getRotation();
        return new double[] {
            pose.getX(),
            pose.getY(),
            pose.getZ(),
            rot.getQuaternion().getW(),
            rot.getQuaternion().getX(),
            rot.getQuaternion().getY(),
            rot.getQuaternion().getZ()
        };
    }
    public static double[] toPoseArray3d(List<Pose3d> poses) {
        double[] result = new double[7 * poses.size()];
        for(int i = 0; i < poses.size(); i++) {
            double[] vals = toPoseArray3d(poses.get(i));
            for(int j = 0; j < vals.length; j++) {
                result[7*i + j] = vals[j];
            }
        }
        return result;
    }
}

File: src/main/kotlin/com/team4099/lib/vision/LogUtil.kt
package com.team4099.lib.vision

import edu.wpi.first.math.geometry.Pose2d
import edu.wpi.first.math.geometry.Pose3d

object LogUtil {
  fun toPoseArray2d(pose: Pose2d): DoubleArray {
    return doubleArrayOf(
      pose.x,
      pose.y,
      pose.rotation.radians
    )
  }

  fun toPoseArray2d(poses: List<Pose2d>): DoubleArray {
    val result = DoubleArray(3 * poses.size)
    for (i in poses.indices) {
      val vals = toPoseArray2d(poses[i])
      for (j in vals.indices) {
        result[3 * i + j] = vals[j]
      }
    }
    return result
  }

  fun toPoseArray3d(pose: Pose3d): DoubleArray {
    val rot = pose.rotation
    return doubleArrayOf(
      pose.x,
      pose.y,
      pose.z,
      rot.quaternion.w,
      rot.quaternion.x,
      rot.quaternion.y,
      rot.quaternion.z
    )
  }

  @JvmStatic
  fun toPoseArray3d(poses: List<Pose3d>): DoubleArray {
    val result = DoubleArray(7 * poses.size)
    for (i in poses.indices) {
      val vals = toPoseArray3d(poses[i])
      for (j in vals.indices) {
        result[7 * i + j] = vals[j]
      }
    }
    return result
  }
}

